Yosemite/Yosemite/Stores/StatsStoreV4.swift

```diff
@@ -285,6 +296,100 @@ extension StatsStoreV4 {
+            onCompletion(StatsStoreV4Error.missingTopProducts)
+            return
+        }
+
+        // Make sure we have all the necesary product data before converting and storing top earners.
+        loadProducts(for: topProducts, siteID: siteID) { [weak self] topProductsResult in
+            guard let self = self else { return }
+
+            switch topProductsResult {
+            case .success(let products):
+                self.mergeAndStoreTopProductsAndStoredProductsIntoTopEarners(granularity: granularity,
+                                                                             date: date,
+                                                                             topProducts: topProducts,
+                                                                             storedProducts: products)
+                onCompletion(nil)
+
+            case .failure(let error):
+                onCompletion(error)
+            }
+        }
+    }
+
+    /// Loads product objects that relates to the top products on a `leaderboard`
+    /// If product objects can't be found in the storage layer, they will be fetched from the remote layer.
+    ///
+    func loadProducts(for topProducts: Leaderboard, siteID: Int64, completion: @escaping (Result<[Product], Error>) -> Void) {
+
+        // Workout if we have stored all products that relate to the given leaderboard
+        let topProductIDs = LeaderboardStatsConverter.topProductsIDs(from: topProducts)
+        let topStoredProducts = loadStoredProducts(siteID: siteID, productIDs: topProductIDs)
+        let missingProductsIDs = LeaderboardStatsConverter.missingProductsIDs(from: topProducts, in: topStoredProducts)
+
+        // Return if we have all the products that we need
+        guard !missingProductsIDs.isEmpty else {
+            completion(.success(topStoredProducts))
+            return
+        }
+
+        // Fetch the products that we have not downloaded and stored yet
+        let productsRemote = ProductsRemote(network: network)
+        productsRemote.loadProducts(for: siteID, by: missingProductsIDs) { result in
+            switch result {
+            case .success(let products):
+                // Return the complete array of products that corresponds to a top product leaderboard
+                let completeTopProducts = products + topStoredProducts
+                completion(.success(completeTopProducts))
+
+            case .failure:
+                completion(result)
+            }
+        }
+    }
+
+    /// Returns all stored products for a given site ID
+    ///
+    func loadStoredProducts(siteID: Int64, productIDs: [Int64] ) -> [Networking.Product] {
+        let products = storageManager.viewStorage.loadProducts(siteID: siteID, productsIDs: productIDs)
+        return products.map { $0.toReadOnly() }
+    }
+
+    /// Merges and stored a top-product leaderboard with an array of stored products into  a `StatsTopEarner` object
```

Comment:
super nit: "Merges and stored" -> "Merges and stores" and "StatsTopEarner" -> "TopEarnerStats"?

Reply:
lol I think I'm dyslexic! 🤪